Update driver

yashika-batra opened the pull request:
changes to driver file: 
- collect everything into 1 file
- move certain functions out (radio and logging functions are in the misc repository)
- pending refactoring of state machine: will make a PR once this is done

future updates:
- pending updates to the reinit function
- pending hardware checks (throwing a custom exception if hardware is accessed without being initialized)

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -54,47 +57,86 @@ def __init__(self):
+        self._vbatt = analogio.AnalogIn(board.BATTERY)  # Define battery voltage
+
+        # Define and initialize hardware
+        self.__init_i2c__()
```
This is a style thing but I think the convention is that functions starting and ending with two underscores `__function__` are python built in functions that have "special" meaning, such as being the object constructor. Since these are not built in functions (unlike `__init__`), these should all start with a single underscore and end with no underscores (`_init_i2c()`). 

See: https://dbader.org/blog/meaning-of-underscores-in-python

thetazero replied:
I like this idea

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -123,136 +172,131 @@ def __init__(self):
+            self.hardware['Coil X'] = True
         except Exception as e:
-            print('[ERROR][H-Bridge U6]', e)
+            print('[ERROR][Initializing H-Bridge U6]', e)
```
Should this be U7?

yashikabatra15 replied:
It should! I must've fixed it in the comments and forgotten to change the print statements

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -123,136 +172,131 @@ def __init__(self):
+            self.hardware['Coil Z'] = True
         except Exception as e:
-            print('[ERROR][H-Bridge U4]', e)
+            print('[ERROR][Initializing H-Bridge U4]', e)
```
U9?

yashikabatra15 replied:
Fixed

thetazero commented:
I really like the look of this. 

But I think before code review/merging it would be nice to sanity check the sensor readings to make sure everything is still hooked up properly. 
Perhaps we should wait for #60 to be a bit more mature in order to do that?

What do you think?

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+    """ return the gyroscope reading from the IMU """
+    return cubesat.IMU.gyro
+
+def temperature():
```
What if we rename this to `temperature_imu()`?

thetazero replied:
I like this, we also don't use this anywhere anyway so we can totally just rename it to whatever we want

yashikabatra15 replied:
I think if we're not using it anywhere currently, it's an easy rename.

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+    if '1' in burn_num:
+        burnwire = cubesat.burnwire1
+    elif '2' in burn_num:
+        return False  # return False because burnwire 2 IC is not set up
```
Should this throw an error?

yashikabatra15 replied:
Ideally, yeah - I'm implementing that with hardware checks / want to put it in the next PR
Right now, it return "False" which indicates that deployment failed (but this case is not handled)

jbwillis replied:
ok

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+
+"""
+IMU-related functions
+TODO: cubesat.imu hardware check
```
Trying to remember the conclusion we came to on this. Did we decide that we just let the system error on `cubesat.imu` not existing?

thetazero replied:
Yes, we were planning for the purposes of this pull request it should just let the system error on cubesat.imu. But later we'd like to to produce the correct error, that being some HardwareNotConnectedErrror or something

jbwillis replied:
Ok, that sounds good.

Since that is the case, and since there is now an issue for this, @yashikabatra15 please remove all of the lines that say `TODO: _____ hardware check`

yashikabatra15 replied:
I'm implementing the error for the next PR, but yeah, for this one we're just going to let the system error. I'll remove all these lines

jbwillis commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
-        # open save file
+
+# initialize Satellite as cubesat
+cubesat = Satellite()
```
This seems a bit weird to me. 

I know we've had some debate about whether to structure this as a full class or as a module. @thetazero how are you thinking this structure would be?

thetazero replied:
It seems we are in some kind of middle ground with initialization. But when it comes to accessing the pycubed library we have exactly what I was thinking we were working towards. 

We need some method to initialize everything, and some way for the functions like `gyro` to access the sensors. Currently we are doing this through a Class, and honestly I think this is fine. Perhaps we should rename the cubesat variable to _cubesat however to make it more clear that it is not meant to be accessed by the end user?

thetazero replied:
I really like this idea however unfortunately the class has to be defined before initialization in python.
For example
```python3
x = Test()
print(x)

class Test:
    def __init__(self):
        self.something='something else'
```
Throws the following error:
```
Traceback (most recent call last):
  File "test.py", line 1, in <module>
    x = Test()
NameError: name 'Test' is not defined
```

However a structure like this does work:
```python3
def doSomethingWithX():
    print(f'This is x: {x}')

class Test:
    def __init__(self):
        self.something='something else'

x = Test()
doSomethingWithX()
```

yashikabatra15 replied:
fixed! I really like these changes

jbwillis commented:
In regards to @thetazero's comment about waiting for the system check script. I'd say let's get started on reviewing both as there will probably be some back and forth between the two.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+Keeping the aliasing so we can do a hardware check for cubesat.radio
+"""
+
+def send(data, *, keep_listening=False, destination=None, node=None,
```
Can we do send=cubesat.radio.send?

jbwillis replied:
Only downside of this is if we change drivers, the function call interface may change. Not likely to happen for the radio, but almost guaranteed for other components like the sun sensors.

thetazero replied:
I think we can keep it this way for radio, because it's just going to conform to the adafruitrfm9x/LoRa standard which we want. But I agree when it comes things like sun sensors.

yashikabatra15 replied:
I think because the only functions that are currently "wrapped" in this way are the radio functions, it makes sense to make this change. If/when we need to introduce more wrapper functions for the other components, we can decide how we want to declare them then

yashikabatra15 replied:
Bad news on the wrapping / reorganization: If I move the Satellite class and initialization to the bottom of the file, it won't let me wrap functions (ex. send = _cubesat.radio.send) because it says it doesn't recognize _cubesat. However, if I access _cubesat within another function it recognizes the name.

I think we might have to leave the function definitions wrapped the way they are

jbwillis replied:
I think that's fine

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+                       destination=destination, node=node, identifier=identifier,
+                       flags=flags)
+
+def listen():
```
listen=cubesat.radio.listen

yashikabatra15 replied:
Forgot python had this quirk; that would save a lot of lines!

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -262,394 +306,168 @@ def reinit(self, dev):
+    cubesat.radio.receive(keep_listening=keep_listening, with_header=with_header,
+                          with_ack=with_ack, timeout=timeout, debug=debug)
+
+def sleep():
```
same wrapping idea sleep=cubesat.radio.sleep

thetazero reviewed (changes requested):
Looks really solid

thetazero commented:
@yashikabatra15 a small change was required in order to get this code not to crash on a pycubed-mini board: https://github.com/PyCubed-Mini/flight_software/pull/82

yashikabatra15 commented:
> In regards to @thetazero's comment about waiting for the system check script. I'd say let's get started on reviewing both as there will probably be some back and forth between the two.

Agreed! I'll upload a new version of this file and also go through the reviews on the check script and let y'all know on Slack.

yashikabatra15 commented:
> @yashikabatra15 a small change was required in order to get this code not to crash on a pycubed-mini board: #82

Saw the change, thanks for catching that!

jbwillis reviewed (changes requested):
This looks great. I commented a couple of minor changes. Biggest change now is updating everything else to use this interface. @thetazero I think it would be good for you to take that on.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +19,182 @@
+
+    for _ in range(50):
+        # 65536 = 2^16, number of increments we can have to voltage
+        vbat += _cubesat._vbatt.value * 3.3 / 65536
```
This is how it was in the original file we were given, my guess is the readings are very noisy. Probably worth investigating later I've created an issue for this: #87

yashikabatra15 replied:
Agreed, I can look into it in more detail later.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +19,182 @@
+    return voltage
+
+
+def fuel_gauge():
```
I agree, it isn't used in the flight application anyway

yashikabatra15 replied:
Would you prefer to make voltage linear between a different value (ex. 3.0) and 4.2? Anything 3.0 or below can return 0 as an indicator that there's an issue with the battery.

thetazero replied:
I'd prefer not to have it

thetazero commented:
@jbwillis I'll update everything to use this interface.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +19,182 @@
+IMU-related functions
+"""
+
+def acceleration():
```
I'd really like to have units in the docstrings for the IMU readings

yashikabatra15 replied:
Will do

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +19,182 @@
+
+def burn(burn_num='1', dutycycle=0, freq=1000, duration=1):
+    """
+    control the burnwire(s)
```
I don't understand what the function does, if you do could you please explain it to me in the docstring.

yashikabatra15 replied:
Yep, will do

thetazero reviewed (changes requested):
Looking really good, only issues I see are doc string related. 

I'm going to go and test this code on the mainboard.

yashikabatra15 commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +19,182 @@
 
+
+"""
+IMU-related functions
```
Will do

yashika-batra commented:
Made the changes you asked for and added the hardware exception code. I also created functions that allow us to access the coil driver hardware and sun sensor hardware. The new PR also updates the coil driver library to a more recent version.

@jbwillis the order that I settled on for the file (based on Losha’s comments regarding python’s constraints) was functions, class definition, _cubesat initialization

I would appreciate it if y'all could take a look at / comment on the TODOs in particular for this review. Namely:
- TODOs regarding exception handling in the coildriver_vout, lux, and burn functions
- TODO regarding whether or not we need the datafields self._stat = {}, self.filenumbers = {}, self.debug = True in the _Satellite class
Regarding the TODOs in the functions that initialize sun sensors - I looked at the schematic and couldn't find I2C bus addresses, so I'll reach out to Neil about the addresses and fix that.

thetazero commented:
@yashika-batra You've made some really great progress, looks pretty much ready for testing and then merging. Few comments/things we need to do
* Please fix the conflict
* I think self._stat/self.filenumbers/self.debug should be removed if they don't cause anything to crash, and removed later otherwise.
* Didn't really look into the burn wire driver update, I think we're best of just testing if it works.
* I made a pull request changing the way we access to decorators instead of the way you currently do it, let me know what you think.
* As for Standby/Coast I really don't know, you're more familiar with the driver at this point, so why don't you make the call here.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+            _cubesat.drv_z.throttle_volts = projected_voltage
+        else:
+            raise HardwareInitException
+    # TODO: we can change how this is handled
```
I think this is fine for now

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+            return _cubesat.sun_xp.lux
+        else:
+            raise HardwareInitException
+    # TODO: we can change how this is handled
```
also fine for now.

thetazero commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -45,611 +233,421 @@ class Satellite:
-    def acceleration(self):
-        """
-        return the accelerometer reading from the IMU
+    def hardwarecheck_device(self, devicestr, device):
```
My bad, I got confused by all the changes, this seems correct.

thetazero commented:
@yashika-batra when running your code on a mainboard I got the following error:
```python
Traceback (most recent call last):
  File "main.py", line 2, in <module>
  File "lib/pycubed.py", line 650, in <module>
  File "lib/pycubed.py", line 246, in __new__
AttributeError: type object '_Satellite' has no attribute 'instance'
```
Did you test your code on a main board?

yashika-batra commented:
> @yashika-batra when running your code on a mainboard I got the following error:
> 
> ```python
> Traceback (most recent call last):
>   File "main.py", line 2, in <module>
>   File "lib/pycubed.py", line 650, in <module>
>   File "lib/pycubed.py", line 246, in __new__
> AttributeError: type object '_Satellite' has no attribute 'instance'
> ```
> 
> Did you test your code on a main board?

Fixed! I messaged you about this on Slack but thought I'd add it here as well for documentation -- I accidentally deleted a line after I tested (and that line initializes the attribute instance). I've since committed this change and tested the driver again and it runs smoothly.

yashika-batra commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -45,611 +233,421 @@ class Satellite:
+            self.hardware['Neopixel'] = True
         except Exception as e:
-            print('[ERROR][RADIO]', e)
+            print('[WARNING][Initializing Neopixel]', e)
```
That's the way it was in the old driver; my best guess is that it's because Neopixel initialization failure isn't fatal to the cubesat's operation, but other initialization failures are. Either way, I can change it to ERROR if that makes more sense

yashika-batra commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+    return voltage
+
+
+def timeon():
```
It's returning just the time. Ideally it should return this time - BOOTTIME, but this is the way it was in the original file. I didn't see this function used anywhere in applications, but it looks like you have the line time.monotonic() - BOOTTIME in multiple places.

I'll implement timeon to return time.monotonic() - BOOTTIME for now, and you can choose if you want to make the refactor.

yashika-batra commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+            # because burnwire2 is not set up, will throw HardwareInitException
+            burnwire = _cubesat.burnwire2
+        else:
+            raise HardwareInitException
```
Agreed - I originally had it print a message and return False and that seemed to make more sense. I'll change it back.

yashika-batra commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+        else:
+            raise HardwareInitException
+    # TODO: we can change how this is handled
+    except HardwareInitException as e:
```
Agreed - I originally had it print a message and return False and that seemed to make more sense. I'll change it back.

yashika-batra commented on state_machine/drivers/pycubedmini/lib/pycubed.py:
```diff
@@ -18,21 +17,210 @@
+        else:
+            raise HardwareInitException
+    # TODO: we can change how this is handled
+    except HardwareInitException as e:
```
Agreed - I originally had it print a message and return False and that seemed to make more sense. I'll change it back.

yashika-batra commented:
Changes:
- Merged new driver refactors, new driver compatibility patch
- Fixed hardware exception TODOs on coil driver, lux, and burn functions (changed them back to print a statement and return a value)
- Fixed the timeon function to return time.monotonic - BOOTTIME (@thetazero as I mentioned in the comments above you can choose if you want to keep that / refactor or not)
- Moved BOOTTIME into the cubesat initialization function - in the case that the driver file is run multiple times when initialized multiple times, we'd be overwriting our BOOTTIME. It makes more sense to maintain it as an attribute of the object itself. I've also created an accessor (BOOTTIME = _cubesat.BOOTTIME) so we don't need to refactor anything for this change.

thetazero commented:
@yashika-batra few things about your changes
* The merges resulted in some errors, I think I cleared them all up in #130 
* As for the TODOs sounds good!
* The fix to the `timeon` functions sounds good 
* Sounds good.
Happy to code review once you confirm that you have resolved the issues I mentioned in #130 and have tested it on a main board.

yashika-batra commented:
> @yashika-batra few things about your changes
> 
> * The merges resulted in some errors, I think I cleared them all up in [Fix Driver Update #130](https://github.com/PyCubed-Mini/flight_software/pull/130)
> * As for the TODOs sounds good!
> * The fix to the `timeon` functions sounds good
> * Sounds good.
>   Happy to code review once you confirm that you have resolved the issues I mentioned in [Fix Driver Update #130](https://github.com/PyCubed-Mini/flight_software/pull/130) and have tested it on a main board.

Merged and tested!

thetazero commented:
Thanks @yashika-batra, I'll look through it

thetazero reviewed (approved):
Great work! 

Really excited to start working with this updated driver.
